File: LeetCode/415. Add Strings.js
// https://leetcode.com/problems/add-strings/description/

const { test } = require('../test');

const addStrings = (num1, num2) => {
  let result = '';

  let rest = 0;

  let index1 = num1.length - 1;
  let index2 = num2.length - 1;

  while (index1 >= 0 || index2 >= 0) {
    const value1 = index1 >= 0 ? Number(num1[index1]) : 0;
    const value2 = index2 >= 0 ? Number(num2[index2]) : 0;

    const currentSum = value1 + value2 + rest;

    rest = Math.trunc(currentSum / 10);

    result = (currentSum % 10) + result;

    index1--;
    index2--;
  }

  return rest ? rest + result : result;
};

test(addStrings, [
  {
    input: ['11', '123'],
    expected: '134',
  },
  {
    input: ['456', '77'],
    expected: '533',
  },
  {
    input: ['0', '0'],
    expected: '0',
  },
]);

File: LeetCode/415. Add Strings.py
# https://leetcode.com/problems/add-strings/description/

from test import test


class Solution:
    def addStrings(self, num1: str, num2: str) -> str:
        result = ""

        index1 = len(num1) - 1
        index2 = len(num2) - 1

        rest = 0

        while index1 >= 0 or index2 >= 0:
            value1 = int(num1[index1]) if index1 >= 0 else 0
            value2 = int(num2[index2]) if index2 >= 0 else 0

            current_sum = value1 + value2 + rest

            rest = current_sum // 10

            result = str(current_sum % 10) + result

            index1 -= 1
            index2 -= 1

        return str(rest) + result if rest else result


solution = Solution()


test(
    solution.addStrings,
    [
        {
            "input": ["11", "123"],
            "expected": "134",
        },
        {
            "input": ["456", "77"],
            "expected": "533",
        },
        {
            "input": ["0", "0"],
            "expected": "0",
        },
    ],
)
